[충남대 FE 김은수] 1단계 React Query 적용

shushuburger opened the pull request:
안녕하세요. 충남대학교 FE 12팀 김은수입니다!!

### 요구사항 및 수행사항
- [x] 본인만의 기준으로 일관된 코드를 작성해주세요.
  - [x] WithApiUi를 통해 조건별 UI 분기에 대해 작성을 해두었기에 해당 컴포넌트를 적용할 수 있는 CategoryContent, RankingContent, ThemeProductList에 대해 수정 작업을 진행했습니다.\
  - [x] WithApiUi에서 T를 기존에는 배열만 지원했다면 단일 객체 역시 지원하게 하여 ThemeHero.tsx에도 WithApiUi를 적용하여주었습니다.
- [x] 기능 단위로 나누어 커밋을 해주세요.
- [x] 기존에 작성했던 API를 React Query를 이용해서 리팩터링 해요. (GET, POST 모두)
  - [x] npm install을 통해 react-query를 설치했습니다.
  - [x] queryClient.ts를 통해 전역 QueryClient 인스턴스를 생성했습니다.
  - [x] App.tsx에 QueryClientProvider로 전체 구조를 감싸주었습니다.
  - [x] useCategoryThemes에서 useFetch 기반을 useQuery 기반으로 변경해주었습니다. (GET)
  - [x] CategoryContent에서 변경된 구조에 맞게 pending, error 제거 및 data = [] 구조로 단순화 작업을 했습니다.
  - [x] useProductRanking.ts 역시 useFetch -> useQuery로 변경해주었습니다. (GET)
  - [x] RankingGroup, RankingContent를 변경된 hook에 맞게 바꿔주었습니다.
  - [x] useThemeProduct.ts에 useFetch를 이용하던 무한 스크롤을 useInfiniteQuery로 대체하였습니다.
  - [x] ThemeProductList.tsx에서 변경된 구조에 맞게 호출 방식을 수정해주었습니다.
  - [x] useLoginMutation.ts를 통해 useMutation을 사용하여주었습니다. (POST)
  - [x] AuthProvider에서는 login 함수 내부에 mutateAsync 사용을 통해 비동기를 바탕으로 로그인 처리를 해주었습니다.
  - [x] QueryClientProvider, AuthProvider의 호출 순서, 중복 호출 문제를 해결하기 위해 App.tsx, main.tsx의 구조를 수정해주었습니다.
  - [x] useOrderMutation을 생성해서 login과 마찬가지로 useMutation을 사용하도록 해주었습니다. (POST)
  - [x] OrderPage에서 추가된 훅을 사용하며 로직을 분리해주어 조금 더 간결해지도록 해주었습니다.
  - [x] useProductSummary에서 useFetch를 이용해 상품 요약 정보를 가져오던 것을 useQuery로 변경해주었습니다. (GET)
  - [x] useThemeInfo 역시 useQuery를 사용하여 정보를 얻어오도록 변경해주었습니다.

shushuburger committed:
docs(readme): 1단계 요구사항 나열

shushuburger committed:
feat(query): 카테고리 테마 조회 기능에 React Query 적용

- useCategoryThemes 훅을 useFetch 기반에서 useQuery 기반으로 변경
- CategoryContent 컴포넌트에서 pending/error 분기 제거
- QueryClient 설정 및 App.tsx에 QueryClientProvider 적용
- react-query 사용을 위한 queryClient.ts 파일 생성
- README에 구현 내용 추가 표시

shushuburger committed:
refactor(query): 선물 랭킹 조회 기능을 react-query 기반으로 전환

- useProductRanking 훅을 useQuery 기반으로 리팩터링
- RankingGroup 컴포넌트에서 data, isLoading, isError 구조 분해 처리
- RankingContent 컴포넌트 props를 React Query에 맞게 정리
- 기존 pending, error 분기 제거 및 캐싱 가능한 queryKey 구성 적용
- README에 구현 완료 내용 표시

shushuburger committed:
refactor(query): 테마별 상품 조회 기능을 react-query 기반으로 리팩터링

- useThemeProducts 훅을 useInfiniteQuery로 전환하여 무한 스크롤 처리
- cursor 기반 페이징을 getNextPageParam으로 설정
- ThemeProductList에서 pending, error 제거하고 isLoading, isError 적용
- README에 구현 내용 표시

shushuburger committed:
docs(readme): 구현 완료한 내용 표시

shushuburger committed:
refactor(auth): 로그인 요청을 useMutation 기반으로 리팩토링

- useLoginMutation 훅 생성 및 mutationFn 내부에서 axios POST 요청 수행
- 로그인 성공 시 sessionStorage에 유저 정보 저장 및 setUser 호출
- AuthProvider에서 login 함수 내부에서 mutateAsync 사용으로 로그인 처리
- README에 구현 완료 내용 표시

shushuburger committed:
refactor(app): 전역 Provider 구조를 main.tsx로 이동하여 App 구성 정리

- QueryClientProvider, ThemeProvider, AuthProvider를 main.tsx에서 루트에 적용
- App.tsx에서는 GlobalStyle, Router, ToastContainer만 렌더링하도록 책임 분리
- QueryClientProvider 중복 감싸기 제거로 오류 및 컨텍스트 충돌 방지
- README에 구현 완료 내용 표시

shushuburger committed:
refactor(order): 주문 요청을 useMutation 기반으로 리팩토링

- useOrderMutation 훅 생성하여 주문 API 요청 로직 분리
- OrderPage에서 axios 직접 호출 제거하고 mutateAsync 사용
- README에 구현 완료 내용 표시

shushuburger committed:
refactor(hook): useProductSummary를 react-query 기반으로 변경

- 기존 useFetch + useCallback 조합 제거
- useQuery를 사용해 상품 요약 정보를 캐싱 및 상태 관리
- isLoading 상태는 query.isPending으로 매핑
- README에 구현 완료 내용 표시

shushuburger committed:
refactor(hook): useThemeInfo를 react-query 기반으로 변경

- useCallback + useFetch 조합 제거
- useQuery를 사용해 테마 정보 비동기 처리 및 상태 관리
- README에 구현 완료 내용 표시

shushuburger committed:
refactor(hook): react-query 도입에 따라 useFetch 훅 제거

- 기존 fetchFn 기반 커스텀 훅은 useQuery로 모두 대체됨
- 중복된 상태 관리 로직 제거 (data, pending, error, refetch)
- useThemeInfo, useProductSummary 등은 react-query 기반 훅 사용
- 더이상 사용하지 않는 mock data 삭제

shushuburger committed:
refactor(ui): WithApiUi 적용으로 로딩/에러/빈 상태 분기 제거

- CategoryContent, RankingContent, ThemeProductList에 WithApiUi 적용
- 각 컴포넌트에서 조건문 제거 후 UI 렌더링에만 집중하도록 구조 개선
- loading은 Loading.tsx의 내용 이용
- README에 구현 내용 표시

shushuburger committed:
refactor(theme): ThemeHero에 WithApiUi 적용하여 조건 분기 제거

- themeInfo의 로딩/에러/빈 상태 처리를 WithApiUi로 추상화
- 컴포넌트는 순수 UI 렌더링에만 집중하도록 구조 단순화
- null-safe 접근을 위해 themeInfo?.속성 형태로 참조 변경
- README에 구현 완료 내용 표현

shushuburger committed:
refactor: 코드 점검하며 개선

- 주문 제출 핸들러 handleOrderSubmit.ts 파일로 분리하여 역할 단일화
- OrderPage에서 onSubmit 로직 간결화
- query param 가공 로직 useRankingParams 훅으로 분리하여 관심사 분리
- WithApiUi, useProductSummary 등 기존 공통 컴포넌트/훅 적극 재사용

shushuburger committed:
fix(api-ui): 빈 데이터일 때 무한 로딩 표시되는 문제 수정

- WithApiUi 컴포넌트에 emptyFallback prop 추가
- data가 빈 배열일 경우 loading 대신 emptyFallback 렌더링하도록 수정
- ThemeProductList에서 상품 없음 메시지를 emptyFallback으로 전달

youngjewoo commented on README.md:
```diff
@@ -5,3 +5,29 @@
 ### 요구사항
 
 - [x] 3차 과제 내용 담긴 나의 코드 불러와 업로드하기
+
```
[ change request ] 1. `npm run lint` 시 발생하는 오류 수정

any를 사용하지 말라는 것과, empty interface 사용한 곳에 대한 오류가 있네요.

any를 멀리할수록 TypeScript 와는 가까워집니다..! 타입을 명확하게 지정해주세요.

empty interface 확장은 emotion에 module augmentation을 위해서 어쩔 수 없는 부분 같아요.
// eslint-disable-next-line 주석을 이용해서 disable 하거나, eslint.config.js에서 설정을 꺼주세요!

그리고 하나는 AuthContext.ts 파일인데요, 하나의 파일에 context와 기타 등등이 함께 선언돼있어서 발생하는 문제입니다.
React는 기본적으로 fast update라는 기능이 내장돼있고, 한 파일에 하나의 컴포넌트를 지향합니다.

그 가정을 바탕으로 코드 변경사항을 빠르게 반영해주는 기능인데, 한 파일에 여러 개를 선언하면 fast update가 원활하게 동작하지 않아서 잘 코드를 구성했음에도 제대로 화면에 보이지 않는 이슈가 발생할 수도 있어요. 그래서 별도의 파일로 분리해주세요!

shushuburger replied:
### AuthContext에 대해 수행한 작업들입니다.
- [x] createContext -> AuthContext.tsx
- [x] Provider 로직 → AuthProvider.tsx
- [x] useContext 호출 → useAuth.ts로 이동
- [x] 관련 타입 정의 → types/auth.ts로 분리

### lint 관련 에러에 대해 수행한 작업입니다.
- any 문제 해결
- src/emtion.d.ts에 주석을 추가하여 disable하게 설정해주었습니다.

youngjewoo commented on src/components/CategorySection/CategoryContent.tsx:
```diff
@@ -1,42 +1,36 @@
+      error={isError}
+      loading={loading}
+      errorFallback={
+        <EmptyText>{ERROR_MESSAGES.FAILED_TO_LOAD_THEMES}</EmptyText>
```
ERROR_MESSAGE를 따로 분리해서 사용하시는 부분이 좋네요!

youngjewoo commented on src/components/common/WithApiUi.tsx:
```diff
@@ -1,22 +1,31 @@
   children: ReactNode;
 }
 
+const isEmpty = (data: unknown): boolean => {
```
함수 역할이 빈 배열을 판단하고 싶은데 맞을까요..? 맞다면 isEmptyArray 면 더 좋을 거 같고요.

이 컴포넌트의 범용성을 높이려면 emptyPredicate같은 함수도 prop으로 받고, 
입력하지 않은 경우 default로 이 `isEmpty` 함수를 넣어주면 다른 조건을 가진 컴포넌트까지 활용할 수 있을 거 같네요!

shushuburger replied:
- isEmpty -> defaultIsEmpty로 변경 ("비었는지 검사하는 기본적인 함수"라는 의미)
- emptyPredicate를 prop으로 받도록 하고 defaultIsEmpty를 기본값으로 설정

위 2가지 작업을 WithApiUi에 해준 다음 
WithApiUi를 사용하는 4개의 컴포넌트 중 "단일 객체"를 data로 넘기는 ThemeHero에는 
emptyPredicate의 값으로 직접 "객체가 비었는지 판단"하는 코드를 넣어주었습니다.
`emptyPredicate={data => !data || Object.keys(data).length === 0}`

youngjewoo replied:
함수를 매개변수를 넘기는 거는 JS에서 함수가 1급 객체(first class citizen)이기 때문에 가능한 패턴입니다.

JS 언어의 장점이고, 프로그램의 유연성을 높여주는 방식이니까 앞으로 이런 공통 유틸을 만드실 때 많이 활용해보세요!

youngjewoo commented on src/hooks/handlers/handleOrderSubmit.ts:
```diff
@@ -0,0 +1,70 @@
+        state: {
+          from: {
+            pathname: location.pathname,
+            search: location.search,
```
```suggestion
            search: window.location.search,
```

브라우저 환경이므로 가급적 window를 붙여주는 것이 좋습니다!
안 붙여도 아마 전역객체가 window라 돌아가긴 할텐데요.. JS가 꼭 브라우저 환경에서만 실행되는 것은 아니니 명확하게 붙여주는 것이 좋습니다.

혹시 모를 로컬 변수로 덮이는 상황도 방지할 수 있고요. 😌

shushuburger replied:
전혀 신경쓰지 못한 부분이었습니다! 앞으로도 신경쓰며 반영하겠습니다!

youngjewoo commented on src/hooks/useRankingParams.ts:
```diff
@@ -0,0 +1,37 @@
+  const rankType =
+    searchParams.get(PARAM_KEYS.RANK_TYPE) || DEFAULT_VALUES.rankType;
+
+  const setTargetType = (value: string) => {
```
```suggestion
  const setTargetType = (value: keyof typeof PARAM_KEYS) => {
```

로 하면 타입이 더 명확하게 추론돼서 사용하는 동료들이 편리하게 사용할 수 있을 거에요!

[예시]
<img width="554" height="115" alt="image" src="https://github.com/user-attachments/assets/5de7ff46-9fda-4ea4-857e-18330e4ff513" />

그리고 hook으로 분리하니까 훨씬 깔끔하군요!

shushuburger replied:
피드백을 수용하려 관련 파일들을 살펴보던 중 TargetType, RankType을 설정하는 것에 있어서 하드코딩이나 비슷한 코드 발견 등 중복되는 부분들이 있는 것을 인지했습니다! 이에 나중에 관리하기 쉬우려면 한 번 중복을 모아주는 시간이 필요하다고 생각했고 "타입 명확화 & 중복 코드 제거"를 목표로 삼아 피드백을 반영해보았습니다!

  - [x] rankingParams.ts를 만들어서 Ranking 부분과 관련있는 하드코딩된 부분과 타입을 상수처리해주었습니다.
  - [x] RankingSort, RankingFilter에서 rankingParams에 정의된 상수를 가져와 사용하도록 해주었습니다.
  - [x] useRankingParams.ts에서도 타입 정의 과정을 수행했습니다.

위의 과정을 통해 rank와 관련된 변수들을 관리하기 쉽고 타입을 명확하게 추론하도록 해주었습니다!

youngjewoo replied:
척하면 척이군요.. 고생하셨습니다. 

앞으로 작업이 훨씬 수월해지실 거에요!

youngjewoo commented on src/hooks/useThemeProducts.ts:
```diff
@@ -1,59 +1,60 @@
-const LIMIT = 10;
+import { useRef } from 'react';
+
+const THEME_PRODUCT_PAGE_LIMIT = 10;
```
오 변수 이름을 좀 더 명확하게 하신게 좋네요.

이런 디테일들이 good 프로그래머와 bad 프로그래머의 차이죠..

youngjewoo commented on src/hooks/useThemeProducts.ts:
```diff
@@ -1,59 +1,60 @@
+    useInfiniteQuery({
+      queryKey: ['themeProducts', themeId],
+      queryFn: ({ pageParam }) =>
+        fetchThemeProducts({ pageParam, themeId: themeId! }),
```
이왕이면 non-null assertion 보다는 themeId가 없는 경우는 명시적으로 처리해주는 것이 좋습니다!

themeId가 없으면 아예 API 호출을 하지 않도록 하는 방법도 있겠네요.

shushuburger replied:
```
      enabled: !!themeId,
      queryFn: ({ pageParam }) => {
        if (!themeId) throw new Error('themeId is required');
        return fetchThemeProducts({ pageParam, themeId });
      },
```

위와 같이 코드를 수정해서 아래 2가지 조건을 만족하도록 변경해주었습니다.
- 명시적으로 themeId가 없는 경우 처리
- themeId가 없으면 API 호출을 하지 않도록 처리

youngjewoo reviewed (changes requested):
오 react-query로 잘 바꿔주셨네요.

빌드 에러나 린트 에러를 제외하고는 크게 수정할 부분이 없을 거 같습니다.

여기까지 오시면서 과제 요구 사항들이 완벽하게 달성되신 분들이 많이 없었던 거 같은데,
이전 과제 내용들 훌륭하게 잘 수행해주셨네요.

빌드, 린드 에러만 수정해서 다시 리뷰 요청 부탁드립니다!

shushuburger committed:
refactor(ui): WithApiUi 적용 및 emptyPredicate 안전성 개선

- CategoryContent: emptyFallback 추가로 빈 목록 대응
- RankingContent: emptyFallback 처리
- ThemeHero:단일 객체를 data로 넘기기에 emptyPredicate prop 이용하여 빈 객체 대응
- ThemeProductList: products 렌더링 전 null 체크로 런타임 오류 방지
- WithApiUi:
  - emptyPredicate props 추가
  - isEmpty -> defaultIsEmpty로 변경
- README에 구현 내용 표시

shushuburger committed:
refactor(order): navigate state에 window.location 명시로 전역 객체 접근 안전성 강화

- location.pathname, location.search 대신 window.location 사용
- 전역 객체 shadowing 및 비브라우저 환경에서의 안정성 확보

shushuburger committed:
refactor(ranking): 필터/정렬 유니언 타입 및 메타데이터 기반 구조로 통합

- rankingParams.ts 생성:
  - TargetType, RankType 유니언 타입을 메타데이터 객체의 key로 자동 정의
  - RANK_TYPE_META, TARGET_TYPE_META에 label/emoji 등 렌더링 정보 포함
  - DEFAULT_VALUES 및 값 목록 상수도 함께 관리

- SegmentedControl 제네릭화:
  - props에 타입 유추 지원
  - 선택값 및 이벤트 타입을 도메인 타입과 연결

- RankingSort/RankingFilter:
  - value/label 하드코딩 제거
  - 메타데이터 기반 options 동적 생성 방식으로 리팩토링
  - selectedSort/Filter 및 onSelect에 RankType/TargetType 적용

- useRankingParams:
  - DEFAULT_VALUES 및 타입 정의를 rankingParams에서 가져오도록 정리

shushuburger committed:
refactor(theme): useThemeProducts에서 themeId non-null assertion 제거

- themeId가 없을 경우 queryFn에서 명시적으로 예외 처리
- enabled 조건을 통해 queryFn 실행 여부를 제어하고 안전성 확보
- themeId! 대신 내부 분기로 오류 발생 시점 명확히 함

shushuburger committed:
refactor(auth): AuthContext 관련 로직 파일 분리 및 구조 개선

- AuthContext 내부에 정의된 context, provider, hook을 분리하여 응집도 향상
- createContext → AuthContext.tsx
- Provider 로직 → AuthProvider.tsx
- useContext 호출 → useAuth.ts로 이동
- 관련 타입 정의 → types/auth.ts로 분리
- main.tsx에서 AuthProvider import 경로 수정
- 모든 파일들의 import 경로 수정

shushuburger committed:
fix(types): Emotion 테마 확장 시 lint 오류 예외 처리

- Theme interface가 empty object로 간주되어 발생한 @typescript-eslint/no-empty-object-type 오류 해결
- 실제 타입 확장을 위한 interface임을 명시적으로 처리하기 위해 eslint-disable-next-line 주석 추가

shushuburger committed:
fix(types): any 타입 제거 및 타입 안정성 개선

- ReceiverInputItem.tsx: getReceiverFields 함수의 errors 파라미터를 FieldErrors<{ receivers: Receiver[] }>로 지정
- error.ts: hasErrorMessage, hasAxiosErrorStatus 함수에서 any → unknown 및 명시적 중첩 타입 단언으로 수정
- README에 구현 내용들 정리

shushuburger commented:
피드백을 반영하고 npm run build, npm run lint 모두 성공적으로 통과함을 확인했습니다!

lint도 npm run을 통해서 검사가 가능한지 몰랐는데 멘토님께서 알려주셔서 새로이 알게되었습니다.
이 명령어를 잘 이용한다면 타입 안전성을 확보하기 훨씬 쉬워질 것 같다는 생각이 들었습니다!
지금까지는 일일이 눌러보며 타입에 관련된 코드를 찾아보고 있었어서 효율이 떨어진다는 생각이었는데 좋은 도구를 알게된 것 같습니다.

### 수행사항
- [x] 피드백 반영
  - [x] WithApiUi의 isEmpty -> defaultIsEmpty로 변경 후 emptyPredicate의 함수로 지정해주었습니다.
  - [x] WithApiUi를 사용하는 4개의 컴포넌트들에 대해 변경된 구조에 맞는지 한 번 더 점검하여주었습니다.
  - [x] 단일 객체를 data로 넘기는 ThemeHero의 경우 컴포넌트 파일 내에서 prop를 통해 대응하도록 나머지 3개는 배열을 data로 넘기기에 defaultIsEmpty를 사용하도록 해주었습니다.
  - [x] handleOrderSubmit.ts에 navigate 부분에 'window'를 추가해주었습니다.
  - [x] rankingParams.ts를 만들어서 Ranking 부분과 관련있는 하드코딩된 부분과 타입을 상수처리해주었습니다.
  - [x] RankingSort, RankingFilter에서 rankingParams에 정의된 상수를 가져와 사용하도록 해주었습니다.
  - [x] useRankingParams.ts에서도 타입 정의 과정을 수행했습니다.
  - [x] themeId가 없는 경우를 명시적으로 처리하도록 수정해주었습니다.
  - [x] AuthContext 파일 세분화 작업을 진행했습니다.
    - [x] createContext -> AuthContext.tsx
    - [x] Provider 로직 → AuthProvider.tsx
    - [x] useContext 호출 → useAuth.ts로 이동
    - [x] 관련 타입 정의 → types/auth.ts로 분리
  - [x] lint 관련 에러 해결
- [x] npm run build, npm run lint 성공적 동작 확인

youngjewoo reviewed (approved):
오우 제가 수정 요청한 핵심을 잘 파악하고 반영해주셨네요!

이번 과제는 문제 없이 동작해서 병합하도록 하겠습니다.
고생 많으셨습니다~~!

youngjewoo commented:
아 이건 디테일인데요. 메인 페이지에서 상품이 있을 때와 없을 때 스크롤이 나타났다가 사라지면서 화면이 덜컥 거리는 현상이 있네요.
이 걸 'scroll jump'라고도 하는데요.. 여유가 되신다면 어떻게 개선하면 좋을지 생각해보고, 개선해보시는 것을 추천드립니다!

![scroll_jump](https://github.com/user-attachments/assets/99f74ede-0f40-485d-9fd4-2b0e5fc1f48f)
